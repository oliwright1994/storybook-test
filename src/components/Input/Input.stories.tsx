import { Stack } from '@chakra-ui/core'
import { Meta } from '@storybook/react/types-6-0'
import { withDesign } from 'storybook-addon-designs'

import React from 'react'

import Input, { IInput } from './Input'

export default { component: Input, title: 'Components/Input', decorators: [withDesign], parameters: {
  componentSubtitle: 'Used to collect information from the used',
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/p9KNWMib8ueXkYnCmMLwsZ/DS-V1.2-Smart?node-id=1%3A12231',
  },
}, } as Meta
export const Standard = (args: IInput): JSX.Element => (
  <Stack>
    <Input {...args} />
  </Stack>
)
Standard.args = {
  errorMessage: 'This value is invalid.',
  isDisabled: false,
  isInvalid: false,
  isReadOnly: false,
  isRequired: false,
  isFullWidth: false,
  placeholder: 'basic',
  label: 'Pot value',
  id: 'amount',
  description: 'Minimum investment: Â£5,000.00',
}
