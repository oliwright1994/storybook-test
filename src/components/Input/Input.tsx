import {
  FormControl,
  FormErrorMessage,
  FormHelperText,
  FormLabel,
  IInput as IChakraInput,
  Input as ChakraInput,
} from '@chakra-ui/core'
import React, { FC, FormEvent } from 'react'

import { Space } from '../../constants'
import Stack from '../Stack'

export interface IInput
  extends Pick<
    IChakraInput,
    'aria-label' | 'aria-describedby' | 'isDisabled' | 'isInvalid' | 'isRequired' | 'isFullWidth' | 'isReadOnly'
  > {
  id: string
  label?: string
  description?: string
  errorMessage?: string
  placeholder?: string
  value?: string
  onChange: (event: FormEvent<HTMLInputElement>) => void
}

const Input: FC<IInput> = ({
  id,
  label,
  description,
  errorMessage,
  isInvalid,
  isRequired,
  isFullWidth = false,
  ...props
}) => {
  return (
    <FormControl isInvalid={isInvalid} isRequired={isRequired}>
      <Stack space={Space.xxs}>
        {label && (
          <FormLabel htmlFor={id} padding={0} verticalAlign="baseline">
            {label}
          </FormLabel>
        )}
        {description && <FormHelperText margin={0}>{description}</FormHelperText>}
        <ChakraInput
          {...props}
          id={id}
          isInvalid={isInvalid}
          isFullWidth={isFullWidth}
          borderColor="gray.400"
          _hover={{ borderColor: 'gray.500' }}
        />
        {errorMessage && <FormErrorMessage>{errorMessage}</FormErrorMessage>}
      </Stack>
    </FormControl>
  )
}

export default Input
