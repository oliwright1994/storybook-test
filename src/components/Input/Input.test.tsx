import { fireEvent } from '@testing-library/react'
import React from 'react'

import { renderWithBrightTheme } from '../../../testing/helpers/themeHelper'
import Input, { IInput } from './Input'

const props: IInput = {
  id: 'input',
  'aria-label': 'aria label',
  'aria-describedby': 'aria describedby',
  isDisabled: false,
  isInvalid: false,
  isRequired: false,
  isFullWidth: false,
  isReadOnly: false,
  onChange: () => null,
}

describe('<Input />', () => {
  it('should render without errors', () => {
    const { asFragment } = renderWithBrightTheme(<Input {...props} label="label" />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should show label, description and placeholder', () => {
    const { getByPlaceholderText, getByText } = renderWithBrightTheme(
      <Input
        {...props}
        label="this is the label"
        description="this is the description"
        placeholder="this is the placeholder"
      />
    )
    expect(getByText(/this is the label/i)).toBeTruthy()
    expect(getByText(/this is the description/i)).toBeTruthy()
    expect(getByPlaceholderText(/this is the placeholder/i)).toBeTruthy()
  })

  it('should not be possible to focus on disabled', () => {
    const { getByDisplayValue } = renderWithBrightTheme(<Input {...props} isDisabled value="this is the value" />)
    expect(getByDisplayValue(/this is the value/i)).toBeDisabled()
  })

  it('should show error if invalid', () => {
    const { getByText } = renderWithBrightTheme(<Input {...props} isInvalid errorMessage="this is the error message" />)
    expect(getByText(/this is the error message/i)).toBeTruthy()
  })

  it('should show asterisk if required', () => {
    const { getByText } = renderWithBrightTheme(<Input {...props} isRequired label="this is the label" />)
    expect(getByText(/this is the label*/i)).toBeTruthy()
  })

  it('should show the initial value, and do callback once when changed', () => {
    const changeHandler = jest.fn()
    const { getByDisplayValue } = renderWithBrightTheme(
      <Input {...props} onChange={changeHandler} value="this is the value" />
    )
    const input = getByDisplayValue(/this is the value/i)
    expect(input).toBeTruthy()
    fireEvent.change(input, { target: { value: 'this is the new value' } })
    expect(changeHandler).toHaveBeenCalledTimes(1)
  })
})
