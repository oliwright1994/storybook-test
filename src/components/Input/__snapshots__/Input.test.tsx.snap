// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="css-0"
    role="group"
  >
    <div
      class="css-0"
    >
      <li
        class="css-1lxwves"
      >
        <label
          class="css-hny2kg"
          for="input"
        >
          label
        </label>
      </li>
      <li
        class="css-qzdso"
      />
      <li
        class="css-qzdso"
      >
        <input
          aria-describedby="aria describedby"
          aria-disabled="false"
          aria-invalid="false"
          aria-label="aria label"
          aria-readonly="false"
          aria-required="false"
          class="css-1az41tb"
          id="input"
        />
      </li>
      <li
        class="css-qzdso"
      />
    </div>
  </div>
</DocumentFragment>
`;
